File: backend/src/main/java/com/pie/backend/service/BpmnXmlParser.java
package com.pie.backend.service;

import com.pie.shared.dto.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import java.io.StringReader;
import java.util.*;

@Service
public class BpmnXmlParser {

    private static final Logger log = LoggerFactory.getLogger(BpmnXmlParser.class);

    public record BpmnParseResult(
            CanonicalProcessGraph graph,
            ProcessKnowledgeDTO knowledge,
            String processName,
            String processId
    ) {}

    public BpmnParseResult parse(String xmlContent) throws Exception {
        if (xmlContent == null || xmlContent.isBlank()) {
            throw new IllegalArgumentException("BPMN XML content cannot be empty");
        }

        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        factory.setNamespaceAware(true);
        DocumentBuilder builder = factory.newDocumentBuilder();
        Document doc = builder.parse(new InputSource(new StringReader(xmlContent)));

        Map<String, GraphNode> nodeMap = new LinkedHashMap<>();
        List<GraphEdge> edges = new ArrayList<>();
        Map<String, String> nodeToLaneMap = new HashMap<>();
        List<String> roleNames = new ArrayList<>();

        // 1. Parse Collaboration and Lanes (Roles & Ownership)
        NodeList laneNodes = doc.getElementsByTagNameNS("*", "lane");
        for (int i = 0; i < laneNodes.getLength(); i++) {
            Element laneElem = (Element) laneNodes.item(i);
            String laneId = laneElem.getAttribute("id");
            String laneName = laneElem.getAttribute("name");
            if (laneName == null || laneName.isBlank()) {
                laneName = "Role " + (i + 1);
            }
            roleNames.add(laneName);

            String roleNodeId = "role-" + slugify(laneName);
            if (!nodeMap.containsKey(roleNodeId)) {
                nodeMap.put(roleNodeId, GraphNode.builder()
                        .id(roleNodeId)
                        .type(NodeType.Role)
                        .label(laneName)
                        .metadata(NodeMetadata.builder().build())
                        .build());
            }

            // Extract flowNodeRef children
            NodeList flowRefs = laneElem.getElementsByTagNameNS("*", "flowNodeRef");
            for (int j = 0; j < flowRefs.getLength(); j++) {
                String refId = flowRefs.item(j).getTextContent().trim();
                if (!refId.isBlank()) {
                    nodeToLaneMap.put(refId, roleNodeId);
                }
            }
        }

        // 2. Identify Primary Process
        NodeList processNodes = doc.getElementsByTagNameNS("*", "process");
        String processId = "imported-bpmn";
        String processName = "Imported BPMN Process";

        if (processNodes.getLength() > 0) {
            Element procElem = (Element) processNodes.item(0);
            if (procElem.hasAttribute("id") && !procElem.getAttribute("id").isBlank()) {
                processId = procElem.getAttribute("id");
            }
            if (procElem.hasAttribute("name") && !procElem.getAttribute("name").isBlank()) {
                processName = procElem.getAttribute("name");
            }
        }

        // 3. Parse Start Events
        NodeList startEvents = doc.getElementsByTagNameNS("*", "startEvent");
        for (int i = 0; i < startEvents.getLength(); i++) {
            Element elem = (Element) startEvents.item(i);
            String id = elem.getAttribute("id");
            String name = elem.getAttribute("name");
            if (name == null || name.isBlank()) name = "Start";

            nodeMap.put(id, GraphNode.builder()
                    .id(id)
                    .type(NodeType.Event)
                    .label(name)
                    .metadata(NodeMetadata.builder().eventType(EventType.start).build())
                    .build());
        }

        // 4. Parse End Events
        NodeList endEvents = doc.getElementsByTagNameNS("*", "endEvent");
        for (int i = 0; i < endEvents.getLength(); i++) {
            Element elem = (Element) endEvents.item(i);
            String id = elem.getAttribute("id");
            String name = elem.getAttribute("name");
            if (name == null || name.isBlank()) name = "End";

            nodeMap.put(id, GraphNode.builder()
                    .id(id)
                    .type(NodeType.Event)
                    .label(name)
                    .metadata(NodeMetadata.builder().eventType(EventType.end).build())
                    .build());
        }

        // 5. Parse Intermediate Events
        parseIntermediateEvents(doc, "intermediateCatchEvent", nodeMap);
        parseIntermediateEvents(doc, "intermediateThrowEvent", nodeMap);

        // 6. Parse Tasks (task, userTask, serviceTask, manualTask, scriptTask, sendTask, receiveTask)
        String[] taskTags = {"task", "userTask", "serviceTask", "manualTask", "scriptTask", "sendTask", "receiveTask"};
        for (String tag : taskTags) {
            NodeList tasks = doc.getElementsByTagNameNS("*", tag);
            for (int i = 0; i < tasks.getLength(); i++) {
                Element elem = (Element) tasks.item(i);
                String id = elem.getAttribute("id");
                String name = elem.getAttribute("name");
                if (name == null || name.isBlank()) name = "Unnamed Activity";

                String roleRef = nodeToLaneMap.get(id);
                NodeMetadata meta = NodeMetadata.builder()
                        .roleRef(roleRef)
                        .build();

                nodeMap.put(id, GraphNode.builder()
                        .id(id)
                        .type(NodeType.Activity)
                        .label(name)
                        .metadata(meta)
                        .build());
            }
        }

        // 7. Parse Gateways (exclusiveGateway, parallelGateway, inclusiveGateway)
        NodeList exGateways = doc.getElementsByTagNameNS("*", "exclusiveGateway");
        for (int i = 0; i < exGateways.getLength(); i++) {
            Element elem = (Element) exGateways.item(i);
            String id = elem.getAttribute("id");
            String name = elem.getAttribute("name");
            nodeMap.put(id, GraphNode.builder()
                    .id(id)
                    .type(NodeType.Gateway)
                    .label(name != null ? name : "")
                    .metadata(NodeMetadata.builder().gatewayType(GatewayType.exclusive).build())
                    .build());
        }

        NodeList paGateways = doc.getElementsByTagNameNS("*", "parallelGateway");
        for (int i = 0; i < paGateways.getLength(); i++) {
            Element elem = (Element) paGateways.item(i);
            String id = elem.getAttribute("id");
            String name = elem.getAttribute("name");
            if (name == null || name.isBlank()) name = "Parallel Gateway";
            nodeMap.put(id, GraphNode.builder()
                    .id(id)
                    .type(NodeType.Gateway)
                    .label(name)
                    .metadata(NodeMetadata.builder().gatewayType(GatewayType.parallel).build())
                    .build());
        }

        NodeList inGateways = doc.getElementsByTagNameNS("*", "inclusiveGateway");
        for (int i = 0; i < inGateways.getLength(); i++) {
            Element elem = (Element) inGateways.item(i);
            String id = elem.getAttribute("id");
            String name = elem.getAttribute("name");
            if (name == null || name.isBlank()) name = "Inclusive Gateway";
            nodeMap.put(id, GraphNode.builder()
                    .id(id)
                    .type(NodeType.Gateway)
                    .label(name)
                    .metadata(NodeMetadata.builder().gatewayType(GatewayType.inclusive).build())
                    .build());
        }

        // 8. Parse Sequence Flows
        NodeList flows = doc.getElementsByTagNameNS("*", "sequenceFlow");
        for (int i = 0; i < flows.getLength(); i++) {
            Element elem = (Element) flows.item(i);
            String id = elem.getAttribute("id");
            String sourceRef = elem.getAttribute("sourceRef");
            String targetRef = elem.getAttribute("targetRef");
            String name = elem.getAttribute("name");

            if (sourceRef != null && targetRef != null && !sourceRef.isBlank() && !targetRef.isBlank()) {
                EdgeType edgeType = (name != null && !name.isBlank()) ? EdgeType.conditional : EdgeType.sequence;
                edges.add(GraphEdge.builder()
                        .id(id != null && !id.isBlank() ? id : "flow-" + (i + 1))
                        .from(sourceRef)
                        .to(targetRef)
                        .edgeType(edgeType)
                        .label(name != null && !name.isBlank() ? name : null)
                        .build());
            }
        }

        // 9. Link Roles to Activities via Association Edges
        for (Map.Entry<String, String> entry : nodeToLaneMap.entrySet()) {
            String activityId = entry.getKey();
            String roleId = entry.getValue();
            if (nodeMap.containsKey(activityId) && nodeMap.containsKey(roleId)) {
                edges.add(GraphEdge.builder()
                        .id("assoc-" + roleId + "-" + activityId)
                        .from(roleId)
                        .to(activityId)
                        .edgeType(EdgeType.association)
                        .label("Performs")
                        .build());
            }
        }

        CanonicalProcessGraph graph = CanonicalProcessGraph.builder()
                .graphId(processId)
                .addNodes(new ArrayList<>(nodeMap.values()))
                .addEdges(edges)
                .build();

        // 10. Synthesize ProcessKnowledgeDTO
        List<String> activities = nodeMap.values().stream()
                .filter(n -> n.getType() == NodeType.Activity)
                .map(GraphNode::getLabel)
                .toList();

        List<String> gateways = nodeMap.values().stream()
                .filter(n -> n.getType() == NodeType.Gateway)
                .map(GraphNode::getLabel)
                .filter(l -> l != null && !l.isBlank())
                .toList();

        List<String> events = nodeMap.values().stream()
                .filter(n -> n.getType() == NodeType.Event)
                .map(GraphNode::getLabel)
                .toList();

        ProcessKnowledgeDTO knowledge = new ProcessKnowledgeDTO(
                activities,
                roleNames,
                roleNames,
                List.of(),
                events,
                gateways,
                List.of("Source BPMN 2.0 Document"),
                List.of("Executed Process Instance"),
                List.of("Process conforms to BPMN 2.0 execution semantics"),
                List.of(),
                List.of()
        );

        log.info("Successfully parsed BPMN XML [{}]: {} nodes, {} edges", processId, nodeMap.size(), edges.size());

        return new BpmnParseResult(graph, knowledge, processName, processId);
    }

    private void parseIntermediateEvents(Document doc, String tagName, Map<String, GraphNode> nodeMap) {
        NodeList list = doc.getElementsByTagNameNS("*", tagName);
        for (int i = 0; i < list.getLength(); i++) {
            Element elem = (Element) list.item(i);
            String id = elem.getAttribute("id");
            String name = elem.getAttribute("name");
            if (name == null || name.isBlank()) name = "Intermediate Event";

            nodeMap.put(id, GraphNode.builder()
                    .id(id)
                    .type(NodeType.Event)
                    .label(name)
                    .metadata(NodeMetadata.builder().eventType(EventType.intermediate).build())
                    .build());
        }
    }

    private String slugify(String input) {
        if (input == null) return "element";
        return input.toLowerCase(Locale.ROOT).replaceAll("[^a-z0-9]+", "-").replaceAll("^-|-$", "");
    }
}

File: frontend/src/utils/bpmnXmlParser.ts
import type {
  CanonicalProcessGraph,
  GraphNode,
  GraphEdge,
  ProcessKnowledgeDTO,
  NodeType,
  EdgeType,
  GatewayType,
  EventType,
} from '../../../backend/src/main/java/com/pie/shared/types/dto';

export interface BpmnClientParseResult {
  graph: CanonicalProcessGraph;
  knowledge: ProcessKnowledgeDTO;
  processName: string;
  processId: string;
}

function slugify(text: string): string {
  if (!text) return 'element';
  return text.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-+|-+$/g, '');
}

export function parseBpmnXmlClient(xmlContent: string): BpmnClientParseResult {
  const parser = new DOMParser();
  const doc = parser.parseFromString(xmlContent, 'application/xml');

  const parserError = doc.querySelector('parsererror');
  if (parserError) {
    throw new Error('XML Parsing failed: ' + parserError.textContent);
  }

  const nodeMap = new Map<string, GraphNode>();
  const edges: GraphEdge[] = [];
  const nodeToLaneMap = new Map<string, string>();
  const roleNames: string[] = [];

  // 1. Lanes & Roles
  const lanes = doc.querySelectorAll('lane');
  lanes.forEach((lane, idx) => {
    const laneName = lane.getAttribute('name') || `Role ${idx + 1}`;
    roleNames.push(laneName);
    const roleId = `role-${slugify(laneName)}`;

    if (!nodeMap.has(roleId)) {
      nodeMap.set(roleId, {
        id: roleId,
        type: 'Role' as NodeType,
        label: laneName,
        metadata: {},
      });
    }

    const flowRefs = lane.querySelectorAll('flowNodeRef');
    flowRefs.forEach((ref) => {
      const refText = (ref.textContent || '').trim();
      if (refText) {
        nodeToLaneMap.set(refText, roleId);
      }
    });
  });

  // 2. Process metadata
  const procElem = doc.querySelector('process');
  const processId = procElem?.getAttribute('id') || 'imported-process';
  const processName = procElem?.getAttribute('name') || 'Imported Business Process';

  // 3. Start Events
  doc.querySelectorAll('startEvent').forEach((elem) => {
    const id = elem.getAttribute('id') || 'start';
    const name = elem.getAttribute('name') || 'Start';
    nodeMap.set(id, {
      id,
      type: 'Event' as NodeType,
      label: name,
      metadata: { eventType: 'start' as EventType },
    });
  });

  // 4. End Events
  doc.querySelectorAll('endEvent').forEach((elem) => {
    const id = elem.getAttribute('id') || 'end';
    const name = elem.getAttribute('name') || 'End';
    nodeMap.set(id, {
      id,
      type: 'Event' as NodeType,
      label: name,
      metadata: { eventType: 'end' as EventType },
    });
  });

  // 5. Intermediate Events
  doc.querySelectorAll('intermediateCatchEvent, intermediateThrowEvent').forEach((elem) => {
    const id = elem.getAttribute('id') || `event-${Date.now()}`;
    const name = elem.getAttribute('name') || 'Intermediate Event';
    nodeMap.set(id, {
      id,
      type: 'Event' as NodeType,
      label: name,
      metadata: { eventType: 'intermediate' as EventType },
    });
  });

  // 6. Activities / Tasks
  const taskSelectors = 'task, userTask, serviceTask, manualTask, scriptTask, sendTask, receiveTask';
  doc.querySelectorAll(taskSelectors).forEach((elem) => {
    const id = elem.getAttribute('id') || `task-${Date.now()}`;
    const name = elem.getAttribute('name') || 'Unnamed Activity';
    const roleRef = nodeToLaneMap.get(id);

    nodeMap.set(id, {
      id,
      type: 'Activity' as NodeType,
      label: name,
      metadata: { roleRef },
    });
  });

  // 7. Gateways (Exclusive, Parallel, Inclusive)
  doc.querySelectorAll('exclusiveGateway').forEach((elem) => {
    const id = elem.getAttribute('id') || `gw-${Date.now()}`;
    const name = elem.getAttribute('name') || '';
    nodeMap.set(id, {
      id,
      type: 'Gateway' as NodeType,
      label: name,
      metadata: { gatewayType: 'exclusive' as GatewayType },
    });
  });

  doc.querySelectorAll('parallelGateway').forEach((elem) => {
    const id = elem.getAttribute('id') || `gw-pa-${Date.now()}`;
    const name = elem.getAttribute('name') || 'Parallel Split / Join';
    nodeMap.set(id, {
      id,
      type: 'Gateway' as NodeType,
      label: name,
      metadata: { gatewayType: 'parallel' as GatewayType },
    });
  });

  doc.querySelectorAll('inclusiveGateway').forEach((elem) => {
    const id = elem.getAttribute('id') || `gw-in-${Date.now()}`;
    const name = elem.getAttribute('name') || 'Inclusive Gateway';
    nodeMap.set(id, {
      id,
      type: 'Gateway' as NodeType,
      label: name,
      metadata: { gatewayType: 'inclusive' as GatewayType },
    });
  });

  // 8. Sequence Flows
  doc.querySelectorAll('sequenceFlow').forEach((elem, idx) => {
    const id = elem.getAttribute('id') || `flow-${idx + 1}`;
    const sourceRef = elem.getAttribute('sourceRef');
    const targetRef = elem.getAttribute('targetRef');
    const name = elem.getAttribute('name');

    if (sourceRef && targetRef) {
      edges.push({
        id,
        from: sourceRef,
        to: targetRef,
        edgeType: name ? ('conditional' as EdgeType) : ('sequence' as EdgeType),
        label: name || null,
      });
    }
  });

  // 9. Link Roles to Activities via Association Edges
  nodeToLaneMap.forEach((roleId, activityId) => {
    if (nodeMap.has(roleId) && nodeMap.has(activityId)) {
      edges.push({
        id: `assoc-${roleId}-${activityId}`,
        from: roleId,
        to: activityId,
        edgeType: 'association' as EdgeType,
        label: 'Performs',
      });
    }
  });

  const graphNodes = Array.from(nodeMap.values());
  const graph: CanonicalProcessGraph = {
    graphId: processId,
    nodes: graphNodes,
    edges,
  };

  const activities = graphNodes.filter((n) => n.type === 'Activity').map((n) => n.label);
  const gateways = graphNodes.filter((n) => n.type === 'Gateway').map((n) => n.label).filter(Boolean);
  const events = graphNodes.filter((n) => n.type === 'Event').map((n) => n.label);

  const knowledge: ProcessKnowledgeDTO = {
    activities,
    actors: roleNames,
    roles: roleNames,
    systems: [],
    events,
    gateways,
    inputs: ['Source BPMN 2.0 XML'],
    outputs: ['Orchestrated Process Execution'],
    businessRules: ['Conforms to OMG BPMN 2.0 specification'],
    risks: [],
    conflicts: [],
  };

  return {
    graph,
    knowledge,
    processName,
    processId,
  };
}
